adapter/internal/discovery/xds/server.go

```diff
@@ -405,23 +412,61 @@ func GenerateEnvoyResoucesForGateway(gatewayName string) ([]types.Resource,
 	}
 
 	envoyGatewayConfig, gwFound := gatewayLabelConfigMap[gatewayName]
-	listener := envoyGatewayConfig.listener
-	if !gwFound || listener == nil {
+	listeners := envoyGatewayConfig.listeners
+	if !gwFound || listeners == nil || len(listeners) == 0 {
 		return nil, nil, nil, nil, nil
 	}
-	routesFromListener := listenerToRouteArrayMap[listener.Name]
-	var vhostToRouteArrayFilteredMap = make(map[string][]*routev3.Route)
-	for vhost, routes := range vhostToRouteArrayMap {
-		if vhost == systemHost || checkRoutes(routes, routesFromListener) {
-			vhostToRouteArrayFilteredMap[vhost] = routes
+	routeConfigs := make([]*routev3.RouteConfiguration, 0)
+	for _, listener := range listeners {
+		for vhost, routes := range vhostToRouteArrayMap {
+			matchedListener, found := common.FindElement(data_holder.GetAllGatewayListeners(), func(listenerLocal gwapiv1b1.Listener) bool {
+				if (listenerLocal.Hostname != nil && common.MatchesHostname(vhost, string(*listenerLocal.Hostname))) {
```

Comment:
can't we do this check only in validation webhooks, because doing this regex operations for each and every update would be very costly

Reply:
This is not a validation check, this is to extract the routes only for the specific vhost and created the routeconfig. In the original code we were creating route config with all the available routes for each route config which is wrong. So in this code I am only extracting the routes that needed for the specific routeconfig